dao/users.js

```diff
@@ -0,0 +1,108 @@
+const debug = require('debug')('calendar-server:reminders');
+const bcrypt = require('bcrypt');
+
+const database = require('./database');
+const { NotFoundError } = require('../utils/errors');
+const {
+  checkPropertyType, checkIsArray
+} = require('../utils/object_validator.js');
+
+function notFoundError(id) {
+  return NotFoundError.createWithSubject('user', { name: 'id', value: id });
+}
+
+const BCRYPT_SALT_ROUNDS = 10;
+const SHA_HASH_SALT = 'todochangeme';
+
+function bcryptHash(plainText) {
+  return new Promise((resolve, reject) => {
+    bcrypt.hash(plainText, BCRYPT_SALT_ROUNDS, function(err, hash) {
+      if (err) {
+        reject(err);
+        return;
+      }
+
+      resolve(hash);
+    });
+  });
+}
+
+function bcryptCompare(plainText, hashed) {
+  return new Promise((resolve, reject) => {
+    bcrypt.compare(plainText, hashed, function(err, result) {
+      if (err) {
+        reject(err);
+        return;
+      }
+
+      resolve(result);
+    });
+  });
+}
+
+module.exports = {
+  create(newUserObject) {
+    checkPropertyType(newUserObject, 'emailAddress', 'string');
+    checkPropertyType(newUserObject, 'password', 'string');
+    checkPropertyType(newUserObject, 'forename', 'string');
+
+    const hashedPasswordPromise = bcryptHash(newUserObject.password);
+
+    return hashedPasswordPromise
+      .then((password) => {
+        const email = newUserObject.emailAddress;
+
+        return database.ready
+          .then((db) => {
+            return db.run(
+              `
+              INSERT INTO
+                user(forename, email, password_hash)
+              VALUES
+                (?, ?, ?)
+              `,
+              newUserObject.forename,
+              email,
+              password
+            );
+          });
+      });
+  },
+  authenticateUser(authObject) {
```

Comment:
needs a small jsdoc to describe how authObject is. Alternatively use ES6 destructuring.